docs/adr/2024-08-14_009-refscripts-fee-change.md

```diff
@@ -0,0 +1,78 @@
+---
+slug: 9
+title: |
+  9. Changes to the fee calculation due to Reference Scripts
+authors: [Alexey Kuleshevich]
+tags: [Accepted]
+---
+
+## Status
+
+Accepted
+
+## Context
+
+It was identified a while ago that there is an overhead associated with deserializing scripts, so much so that it would be possible to create a very large script that was fairly expensive to deserialize, but very cheap to execute. This opened up an attack vector when such a Plutus script would be used as a reference script. This problem was exacerbated by the fact that there was no real limit on the total size of reference scripts that could be used in a transaction, thus being limited only by the size of the transaction itself. Therefore this opened up Cardano to a DDoS attack where an attacker could submit many such transactions that would cost very little, but would be expensive for a `cardano-node` to validate.
+
+In order to prevent such an attack a `"minFeeRefScriptCostPerByte"` protocol parameter was introduced in the Conway era. The idea was fairly simple: we would calculate the total size of reference scripts used by a transaction and multiply it by the value specified by this protocol parameter. Result would be added to the transaction fee. However, in order for this approach to be a definite deterrent of such attacks this parameter would have to be set to a fairly high value. Community was not very keen on having this value set to a high enough value, since that could make reference scripts almost as expensive to use as regular scripts, thus significantly reducing their usability. Taking this fact into consideration, a decision was made to set this parameter to a fairly moderate value to at least deter potential attacks like that by at least making it pretty expensive, instead of prohibitively expensive. This would allow us to fine tune the parameter or potentially change the pricing algorithm at a later point after we were already in the Conway era. This reasoning was coming from the fact that at that point we would be able to reveal to a wider audience the true nature of this issue.
+
+Unfortunately things did not go exactly as planned, because on [June 25th 2024 an attack like this actually took place](https://cardanospot.io/news/ddos-attack-on-the-cardano-blockchain-mRIKAzZTNnzq5NGd). This attack forced us to make a quick decision on implementing a definite prevention of such attacks, while trying to not have a drastic impact on the common use case that DApp developers rely on so much.
```

Comment:
Out of curiosity, what was the size of the reference scripts used in this attack?

Reply:
There were two "waves" of the attack, separated by ~15.5h:

 - First one from slot `127669141` to `127675890`, with 289 transactions, out of which 70 referenced ~1.5MB of ref scripts each, and the rest referenced ~3.1MB of ref scripts each.
 - Second one from slot `127731847` to `127778439`, with 4145 transactions, with ~3.1 MB of ref scripts each (between 3080701B and 3080772B).

Example tx: https://cexplorer.io/tx/d21f6e0309896a1a428a727302d0f66895846e6f0cf8747b9afacbfb0a3e0b51